### PR TITLE
propagate exception

### PR DESCRIPTION
**Pull request by guihao-liang**
The root cause for Linux segfault might be caused by `fin` continues to read and deserialize garbage data when underlying file IO fails.

```
void block_manager::init_segment(std::shared_ptr<block_manager::segment>& seg) {
  ...
  // read 8 bytes
  fin->read(reinterpret_cast<char*>(&footer_size), sizeof(footer_size));

  // deserialize the block information
  fin->clear();
  fin->seekg(filesize - footer_size - sizeof(footer_size), std::ios_base::beg);
  iarchive iarc(*fin);
  iarc >> seg->blocks;

  seg->inited = true;
  seg->file_size = filesize;
}
```

One should always check `fin.good()` after using it. As @ylow mentioned, most designs are based on the assumption that the retry logic (exception handling may involve) happens at very low-level IO operations and those IO will succeed in the end, people tend to not to check `fin` state bits.

This change will enable IO to throw since by default, IO sets its state bits instead of throwing errors.

After discussing with @ylow, I think it's reasonable to propagate exception on `badbit` because those types of errors usually mean irrecoverable.

This is a behavioral change on IO. And it's a good change.

### CODE REVIEW

**Comment by guihao-liang** [comment]
Unexpectedly included some format changes. I hope it won't bother.

The general idea behind is that `general_fstream` wraps `union_fstream`. In order to let exception to be poped, we need to turn the gate open at each wrapping layer.

File: src/core/storage/fileio/general_fstream.cpp [hunk 1]
```diff
@@ -21,27 +22,33 @@ general_ifstream::general_ifstream(std::string filename)
...
-     } catch (const std::exception& e) {
-      log_and_throw_io_failure("Cannot open " + sanitize_url(filename) + " for read. " + e.what());
-    } catch (std::string e) {
-      log_and_throw_io_failure("Cannot open " + sanitize_url(filename) + " for read. " + e);
-    } catch (...) {
-      log_and_throw_io_failure("Cannot open " + sanitize_url(filename));
-    }
+    try : general_ifstream_base(filename),
+          opened_filename(filename) {
+  exceptions(std::ios_base::badbit);
```
Comment (guihao-liang) [inline src/core/storage/fileio/general_fstream.cpp:27]: That's why you may need, http://www.cplusplus.com/reference/ios/ios/exceptions/

badbit includes failbit, anyway.

File: src/core/storage/fileio/union_fstream.cpp [hunk 2]
```diff
@@ -148,34 +153,29 @@ union_fstream::union_fstream(std::string url,
...
 /// Returns the current stream type. Whether it is a HDFS stream or a STD stream
-union_fstream::stream_type union_fstream::get_type() const {
-  return type;
-}
+union_fstream::stream_type union_fstream::get_type() const { return type; }
 
-std::shared_ptr<std::istream> union_fstream::get_istream () {
+std::shared_ptr<std::istream> union_fstream::get_istream() {
   ASSERT_TRUE(input_stream != NULL);
+  input_stream->exceptions(std::ios_base::badbit);
```
Comment (guihao-liang) [inline src/core/storage/fileio/union_fstream.cpp:166]: this line and line 172 is what you need to review.

**Comment by guihao-liang** [comment]
passed all internal tests.

File: src/core/storage/fileio/general_fstream.cpp [hunk 3]
```diff
@@ -1,13 +1,14 @@
 /* Copyright © 2017 Apple Inc. All rights reserved.
  *
  * Use of this source code is governed by a BSD-3-clause license that can
- * be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+ * be found in the LICENSE.txt file or at
```
Comment (TobyRoseman) [inline src/core/storage/fileio/general_fstream.cpp:4]: I'd really prefer not add a new line here. Let's keep all of our copyright headers the same. At some point we might want to right a script to change our headers and I could see something like this breaking that.

File: src/core/storage/fileio/general_fstream.cpp [hunk 4]
```diff
@@ -51,9 +58,7 @@ size_t general_ifstream::get_bytes_read() {
   return this->general_ifstream_base::operator->()->get_bytes_read();
 }
 
-std::string general_ifstream::filename() const {
-  return opened_filename;
-}
+std::string general_ifstream::filename() const { return opened_filename; }
```
Comment (TobyRoseman) [inline src/core/storage/fileio/general_fstream.cpp:61]: Should the return value be `const`?

File: src/core/storage/fileio/union_fstream.cpp [hunk 5]
```diff
@@ -1,9 +1,11 @@
 /* Copyright © 2017 Apple Inc. All rights reserved.
  *
  * Use of this source code is governed by a BSD-3-clause license that can
- * be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+ * be found in the LICENSE.txt file or at
```
Comment (TobyRoseman) [inline src/core/storage/fileio/union_fstream.cpp:4]: Same as my first comment.

File: src/core/storage/fileio/general_fstream.cpp (hunk 3 above)
Comment (guihao-liang) [reply src/core/storage/fileio/general_fstream.cpp:4]: Or we can run clang-format on all of our files to make this exception to be normal. Hoyt had some scripts adding headers and those scripts work on this kind of scenario. Again, if adding headers, a program should be robust for block and line comments.

File: src/core/storage/fileio/general_fstream.cpp (hunk 4 above)
Comment (guihao-liang) [reply src/core/storage/fileio/general_fstream.cpp:61]: it's a copy.

**Pull request merged**